host_applications/linux/apps/raspicam/gl_scenes/shader.c

```diff
@@ -0,0 +1,287 @@
+             glUniformMatrix2fv( array[i].loc, 1, GL_FALSE, array[i].param );
+             break;
+           case GL_FLOAT_MAT3:
+             glUniformMatrix3fv( array[i].loc, 1, GL_FALSE, array[i].param );
+             break;
+           case GL_FLOAT_MAT4:
+             glUniformMatrix4fv( array[i].loc, 1, GL_FALSE, array[i].param );
+             break;
+
+           /* textures */
+           case GL_SAMPLER_2D:
+             glUniform1i(array[i].loc, array[i].param[0]);
+             break;
+           case GL_SAMPLER_CUBE: break;
+             glUniform1i(array[i].loc, (array[i].param[0]));
+             break;
+           default:
+           ;
+         }
+      // remove flag because the value is in GL's state now...
+      array[i].flag=0;  
+
+    }
+  }
+  return 0;
+}
+
+static int shader_redraw(RASPITEX_STATE *state)
+{
+   // Start with a clear screen
+   //~ static float offset = 0.0;
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+    // Bind the OES texture which is used to render the camera preview
+    glBindTexture(GL_TEXTURE_EXTERNAL_OES, state->texture);
+
+    //~ offset += 0.05;
+    GLCHK(glUseProgram(shader.program));
+    GLCHK(glEnableVertexAttribArray(shader.attribute_locations[0]));
+    GLfloat varray[] = {
+        -1.0f, -1.0f,
+        1.0f,  1.0f,
+        1.0f, -1.0f,
+
+        -1.0f,  1.0f,
+        1.0f,  1.0f,
+        -1.0f, -1.0f,
+    };
+    GLCHK(glVertexAttribPointer(shader.attribute_locations[0], 2, GL_FLOAT, GL_FALSE, 0, varray));
+    
+    shader_update_shader(state);
+
+    //GLCHK(glUniform1f(shader.uniform_locations[1], offset));
+    GLCHK(glDrawArrays(GL_TRIANGLES, 0, 6));
+
+    GLCHK(glDisableVertexAttribArray(shader.attribute_locations[0]));
+    GLCHK(glUseProgram(0));
+    return 0;
+}
+
+int shader_open(RASPITEX_STATE *state)
+{
+   const GLchar *defaultVertexShader = 
+    "attribute vec2 vertex;\n"
+    "varying vec2 texcoord;"
+    "void main(void) {\n"
+    "   texcoord = 0.5 * (vertex + 1.0);\n"
+    "   gl_Position = vec4(vertex, 0.0, 1.0);\n"
+    "}\n";
+
+   const GLchar *defaultFragmentShader =  
+      "#extension GL_OES_EGL_image_external : require\n"
+      "varying vec2 texcoord;"
+      "uniform samplerExternalOES tex;"
+      "void main() \n"
+      "{ \n"
+      " gl_FragColor = texture2D(tex, texcoord); \n"
+      "} \n";
+
+   FILE * pfFile=NULL;
+   FILE * pvFile=NULL;
+   long lSize;
+   int result;
+
+   pfFile = fopen (state->fragment_shader_filename,"rb");
+   if (pfFile==NULL) {
+      printf("No or invalid fragment file provided, used the default one\n");
+      shader.fragment_source=defaultFragmentShader;
+   } else {
+
+   // obtain file size:
+   fseek (pfFile , 0 , SEEK_END);
+   lSize = ftell (pfFile);
+   rewind (pfFile);
+
+   // allocate memory to contain the whole file:
+   shader.fragment_source = (GLchar*) malloc (sizeof(GLchar)*lSize);
+   if (shader.fragment_source == NULL) {fputs ("Memory error",stderr); exit (2);}
```

Comment:
If these files are huge then it might be better to use mmap instead of malloc. If they are small then it's simpler to assume that malloc will succeed in userspace linux applications